#include "ingamescene.h"
#include "ui_ingamescene.h"

#include <algorithm>
#include <QDebug>
#include <QTimer>

#include "winscene.h"
#include "losescene.h"

InGameScene::InGameScene(QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::InGameScene)
{
    ui->setupUi(this);
    this->setWindowTitle("InGame");
    this->setFixedSize(800, 600);

    // ä¸€å¼€å§‹å…ˆæŠŠå‡ºç‰ŒæŒ‰é’®éšè—ï¼ˆåªåœ¨å‡ºç‰Œé˜¶æ®µ / ç©å®¶å›åˆæ˜¾ç¤ºï¼‰
    ui->btn_play->hide();
    ui->btn_pass->hide();
    ui->btn_hint->hide();

    // å¼€ä¸€å±€æ¸¸æˆï¼ˆæ´—ç‰Œ + å‘ç‰Œ + è®¾ç½®å«åœ°ä¸»èµ·å§‹ç©å®¶ï¼‰
    initGame();

    // è¿æ¥å«åˆ†æŒ‰é’®
    connect(ui->btn_notcall, &QPushButton::clicked, this, [this]() { onCallScore(0); });
    connect(ui->btn_1p,      &QPushButton::clicked, this, [this]() { onCallScore(1); });
    connect(ui->btn_2p,      &QPushButton::clicked, this, [this]() { onCallScore(2); });
    connect(ui->btn_3p,      &QPushButton::clicked, this, [this]() { onCallScore(3); });

    // è¿æ¥å‡ºç‰ŒåŒºåŸŸæŒ‰é’®
    connect(ui->btn_play, &QPushButton::clicked, this, &InGameScene::onPlayClicked);
    connect(ui->btn_pass, &QPushButton::clicked, this, &InGameScene::onPassClicked);
    connect(ui->btn_hint, &QPushButton::clicked, this, &InGameScene::onHintClicked);
    qDebug() << "æˆ‘æ˜¯æ–°ç‰ˆæœ¬";
}

InGameScene::~InGameScene()
{
    delete ui;
}

// â€”â€” å®Œæ•´å¼€å±€ / é‡å¼€ä¸€å±€ â€”â€”
void InGameScene::initGame()
{
    // æ¸…ç©ºæ—§ UI
    clearLastPlay();
    qDeleteAll(m_handPanels);
    qDeleteAll(m_landlordPanels);
    m_handPanels.clear();
    m_landlordPanels.clear();
    setStatusText("Status: New Game Start");

    // é‡æ–°å¼€å±€ï¼šæ´—ç‰Œ + å‘ç‰Œ + è®¾ç½®å«åœ°ä¸»èµ·å§‹ç©å®¶
    qDebug() << "===== GameStart() æ–°ä¸€å±€ =====";
    m_game.GameStart();

    // æ ¹æ®å½“å‰ game çŠ¶æ€é‡å»º UI
    setupUIForCurrentGame();
}

// â€”â€” åªæ ¹æ®ç°æœ‰ game çŠ¶æ€é‡ç”» UIï¼Œä¸å† GameStart â€”â€”
void InGameScene::setupUIForCurrentGame()
{
    // 1. ç”»å‡ºä¸Šæ–¹ 3 å¼ åœ°ä¸»ç‰Œï¼ˆå…ˆèƒŒé¢ï¼‰
    createLandlordPanels(false);

    // 2. ç”»å‡ºç©å®¶ 0 çš„æ‰‹ç‰Œï¼ˆæ­£é¢ï¼‰
    createPlayer0HandPanels();

    // 3. æ›´æ–° AI å‰©ä½™ç‰Œæ•°ï¼ˆåˆå§‹éƒ½æ˜¯ 17ï¼‰
    updateAiRemainLabels();

    // 4. è®© AI å…ˆå«åˆ°è½®åˆ°ç©å®¶ 0 ä¸ºæ­¢
    while (m_game.GetStatus() == Status::GetLandlord &&
           m_game.GetCurrentPlayer() &&
           m_game.GetCurrentPlayer()->GetId() != 0)
    {
        qDebug() << "[init] AI" << m_game.GetCurrentPlayer()->GetId() << "æ­£åœ¨å«åˆ†...";
        setStatusText("Status: AI calling...");
        m_game.CallLandlordPhase();

        // å¦‚æœ AI å«åˆ†åå·²ç»ç¡®å®šåœ°ä¸»ï¼Œç›´æ¥è¿›å…¥å‘åœ°ä¸»ç‰Œé˜¶æ®µ
        if (m_game.GetStatus() == Status::SendLandlordCard)
        {
            qDebug() << "[init] AI å·²å†³å®šåœ°ä¸»ï¼Œå‘åœ°ä¸»ç‰Œ";
            setStatusText("Status: Game start!");
            m_game.SendLandlordCard();
            createLandlordPanels(true);
            refreshPlayer0HandPanels();
            hideCallButtons();
            enterDiscardPhase();
            return;
        }
    }

    // 5. å¦‚æœè½®åˆ°ç©å®¶ 0ï¼Œæ˜¾ç¤ºå«åˆ†æŒ‰é’®
    if (m_game.GetStatus() == Status::GetLandlord &&
        m_game.GetCurrentPlayer() &&
        m_game.GetCurrentPlayer()->GetId() == 0)
    {
        qDebug() << "[setupUIForCurrentGame] è½®åˆ°ç©å®¶ 0 å«åˆ†";
        setStatusText("Status: You Call!");
        ui->btn_notcall->show();
        ui->btn_1p->show();
        ui->btn_2p->show();
        ui->btn_3p->show();
    }
}
void InGameScene::createLandlordPanels(bool faceUp)
{
    // æ¸…ç©ºæ—§çš„
    qDeleteAll(m_landlordPanels);
    m_landlordPanels.clear();

    int topY    = 60;
    int centerX = width() / 2;
    int spacing = 40;

    for (int i = 0; i < 3; ++i)
    {
        int id = m_game.GetLandlordCard(i);

        auto *p = new CardPanel(this);
        p->setCardId(id);
        p->setFaceUp(faceUp);

        int x = centerX + (i - 1) * spacing - p->width() / 2;
        p->move(x, topY);
        p->show();

        m_landlordPanels.append(p);
    }
}

void InGameScene::createPlayer0HandPanels()
{
    qDeleteAll(m_handPanels);
    m_handPanels.clear();

    Player *human = m_game.GetPlayer(0);
    const std::multiset<int> &cards = human->GetCards();

    std::vector<int> sorted(cards.begin(), cards.end());
    std::sort(sorted.begin(), sorted.end(), [](int a, int b) {
        return CardGroup::Translate(a) < CardGroup::Translate(b);
    });

    int spacing    = 30;
    int leftMargin = 40;
    int baseY      = height() - 150;

    for (int i = 0; i < (int)sorted.size(); ++i)
    {
        int id = sorted[i];

        auto *p = new CardPanel(this);
        p->setCardId(id);
        p->setFaceUp(true);

        p->move(leftMargin + i * spacing, baseY);
        p->show();

        connect(p, &CardPanel::clicked, this, &InGameScene::onCardClicked);

        m_handPanels.append(p);
    }
}

void InGameScene::refreshPlayer0HandPanels()
{
    createPlayer0HandPanels();
}

void InGameScene::hideCallButtons()
{
    ui->btn_notcall->hide();
    ui->btn_1p->hide();
    ui->btn_2p->hide();
    ui->btn_3p->hide();
}
void InGameScene::onCallScore(int score) {
    qDebug() << "[onCallScore] ç©å®¶ç‚¹å‡»å«åˆ†æŒ‰é’®ï¼Œåˆ†æ•°:" << score;

    m_game.PlayerCallLandlord(score);   // åªè°ƒè¿™ä¸€å¥

    hideCallButtons();

    if (m_game.GetStatus() == Status::SendLandlordCard) {
        qDebug() << "-- åœ°ä¸»å·²å†³å®š --";
        m_game.SendLandlordCard();
        createLandlordPanels(true);
        refreshPlayer0HandPanels();
        enterDiscardPhase();
        return;
    }

    if (m_game.GetStatus() == Status::GetLandlord &&
        m_game.GetLandlord() == nullptr)
    {
        qDebug() << "ã€æ— äººå«åœ°ä¸»ã€‘é‡æ–°å¼€å±€";
        initGame();
        return;
    }
}


void InGameScene::enterDiscardPhase()
{
    // ================== å…ˆæ£€æŸ¥æ˜¯å¦å·²ç» GameOver ==================
    if (m_game.GetStatus() == Status::GameOver)
    {
        qDebug() << "æ¸¸æˆç»“æŸï¼Œå¼¹å‡ºç»“æœç•Œé¢";

        Player* human = m_game.GetPlayer(0);
        bool humanWin = (human && human->GetRemain() == 0);

        // ç¦ç”¨ / éšè—å‡ºç‰ŒæŒ‰é’®ï¼Œé˜²æ­¢ç»§ç»­æ“ä½œ
        ui->btn_play->setEnabled(false);
        ui->btn_pass->setEnabled(false);
        ui->btn_hint->setEnabled(false);
        ui->btn_play->hide();
        ui->btn_pass->hide();
        ui->btn_hint->hide();

        if (humanWin) {
            auto *dlg = new WinScene(this);
            dlg->show();
        } else {

            auto *dlg = new LoseScene(this);
            dlg->show();
        }

        return;    // å·²ç»ç»“æŸï¼Œä¸å†è¿›å…¥å‡ºç‰Œé˜¶æ®µ
    }
    // ================== æ­£å¸¸å‡ºç‰Œé˜¶æ®µ ==================
    if (m_game.GetStatus() != Status::Discard)
        return;

    // æ¯æ¬¡è¿›é˜¶æ®µéƒ½åˆ·æ–°ä¸€ä¸‹ AI å‰©ä½™ç‰Œæ•°
    updateAiRemainLabels();

    Player* cur = m_game.GetCurrentPlayer();
    if (!cur) {
        qDebug() << "å½“å‰ç©å®¶ä¸ºç©ºï¼ŒçŠ¶æ€:" << (int)m_game.GetStatus();
        return;
    }

    // ç©å®¶å›åˆ
    if (cur->GetId() == 0)
    {
        qDebug() << "è½®åˆ°ç©å®¶å‡ºç‰Œ";
        setStatusText("Status: Your Turn!");
        ui->btn_play->show();
        ui->btn_pass->show();
        ui->btn_hint->show();
        ui->btn_play->setEnabled(true);
        ui->btn_pass->setEnabled(true);
        ui->btn_hint->setEnabled(true);
        return;
    }

    // AI å›åˆï¼ˆåŠ  2 ç§’å»¶è¿Ÿï¼‰
    qDebug() << "è½®åˆ° AI" << cur->GetId() << " å‡ºç‰Œï¼Œ2 ç§’åæ‰§è¡Œ";
    setStatusText(QString("Statusï¼š AI%1 Playâ€¦").arg(cur->GetId()));
    ui->btn_play->hide();
    ui->btn_pass->hide();
    ui->btn_hint->hide();

    QTimer::singleShot(2000, this, [this]() {
        if (m_game.GetStatus() != Status::Discard)
            return;

        Player* curAI = m_game.GetCurrentPlayer();
        if (!curAI || curAI->GetId() == 0)
        {
            // è¿™ 2 ç§’å†…æœ‰å¯èƒ½æ¸¸æˆç»“æŸæˆ–è½®åˆ°ç©å®¶äº†
            enterDiscardPhase();
            return;
        }

        qDebug() << "AI" << curAI->GetId() << " å¼€å§‹å‡ºç‰Œ";

        clearLastPlay();       // å…ˆæ¸…ç©ºä¸Šä¸€æ¬¡æ˜¾ç¤ºï¼Œé¿å…å åŠ 

        m_game.DiscardPhase(); // è®© Game å†…éƒ¨æŒ‰è§„åˆ™å‡ºç‰Œ / è¿‡ç‰Œï¼ˆå¯èƒ½è¿ç€å‡ ä¸ª AIï¼‰

        showLastPlay();        // åœ¨ä¸­é—´æ˜¾ç¤ºâ€œä¸Šå®¶å‡ºçš„ç‰Œâ€
        updateAiRemainLabels();

        // ç»§ç»­ä¸‹ä¸€ä½ï¼ˆå¯èƒ½åˆæ˜¯ AIï¼Œä¹Ÿå¯èƒ½è½®åˆ°ç©å®¶ï¼‰
        enterDiscardPhase();
    });
}


void InGameScene::onCardClicked()
{
    CardPanel *panel = qobject_cast<CardPanel*>(sender());
    if (!panel) return;

    bool sel = !panel->isSelected();
    panel->setSelected(sel);

    int dy = sel ? -20 : 20;
    panel->move(panel->x(), panel->y() + dy);
}

// â€”â€” UI åˆ é™¤ç©å®¶åˆšåˆšå‡ºæ‰çš„ç‰Œ â€”â€”
void InGameScene::applyPlayerDiscardToUI(const std::vector<int>& indices)
{
    // å¿…é¡»ä»å¤§åˆ°å°åˆ ï¼Œå¦åˆ™å‰é¢çš„åˆ äº†ä¸‹æ ‡ä¼šå˜
    std::vector<int> sorted = indices;
    std::sort(sorted.begin(), sorted.end(), std::greater<int>());

    for (int idx : sorted)
    {
        if (idx < 0 || idx >= m_handPanels.size()) continue;
        CardPanel* p = m_handPanels[idx];
        m_handPanels.remove(idx);
        p->deleteLater();
    }

    // é‡æ–°æ’ç‰ˆå‰©ä½™çš„ç‰Œ
    int spacing    = 30;
    int leftMargin = 40;
    int baseY      = height() - 150;

    for (int i = 0; i < m_handPanels.size(); ++i)
    {
        m_handPanels[i]->move(leftMargin + i * spacing, baseY);
    }
}

// â€”â€” æ¸…ç©ºâ€œä¸Šå®¶å‡ºçš„ç‰Œâ€çš„æ˜¾ç¤º â€”â€”
void InGameScene::clearLastPlay()
{
    qDeleteAll(m_lastPlayPanels);
    m_lastPlayPanels.clear();
}

// â€”â€” æ ¹æ® Game::GetLastDiscard() æŠŠâ€œä¸Šå®¶å‡ºçš„ç‰Œâ€æ˜¾ç¤ºåœ¨ä¸­é—´ â€”â€”
void InGameScene::showLastPlay()
{
    clearLastPlay();

    const CardGroup &grp = m_game.GetLastDiscard();
    Player* last = m_game.GetLastPlayer();

    if (!last) {
        qDebug() << "showLastPlay: lastPlayer ä¸º null";
        return;
    }

    if (grp.GetCount() == 0)
    {
        qDebug() << "showLastPlay: ä¸Šå®¶ç©å®¶ID =" << last->GetId()
                << " å‡ºç‰Œæ•° = 0ï¼ˆå¯èƒ½æ˜¯ passï¼‰";
        return;
    }

    std::vector<int> cards(grp.GetCards().begin(), grp.GetCards().end());
    std::sort(cards.begin(), cards.end(), [](int a, int b){
        return CardGroup::Translate(a) < CardGroup::Translate(b);
    });

    int count   = (int)cards.size();
    int centerX = width() / 2;
    int baseY   = height() / 2 - 80;
    int spacing = 30;

    qDebug() << "åœ¨ä¸­é—´æ˜¾ç¤ºä¸Šå®¶å‡ºç‰Œ, ç©å®¶ID:" << last->GetId()
             << ", å¼ æ•°:" << count;

    for (int i = 0; i < count; ++i)
    {
        CardPanel *p = new CardPanel(this);
        p->setCardId(cards[i]);
        p->setFaceUp(true);

        int x = centerX + (i - count/2.0) * spacing - p->width()/2;
        p->move(x, baseY);
        p->show();

        m_lastPlayPanels.append(p);
    }
}

// â€”â€” æ›´æ–°ä¸¤ä¸ª AI çš„å‰©ä½™ç‰Œæ•°ï¼ˆä½¿ç”¨ä½ åœ¨ .ui é‡Œæ”¾çš„ labelï¼‰ â€”â€”
void InGameScene::updateAiRemainLabels()
{
    if (!ui) return;

    Player* ai1 = m_game.GetPlayer(1);
    Player* ai2 = m_game.GetPlayer(2);
    if (!ai1 || !ai2) return;

    int r1 = ai1->GetRemain();
    int r2 = ai2->GetRemain();

    ui->label_ai1Remain->setText(
        QString("AI1 å‰©ä½™: %1").arg(r1)
        );
    ui->label_ai2Remain->setText(
        QString("AI2 å‰©ä½™: %1").arg(r2)
        );

    qDebug() << "updateAiRemainLabels: AI1 =" << r1 << ", AI2 =" << r2;
}

void InGameScene::onPlayClicked()
{
    // æ”¶é›†å½“å‰é€‰ä¸­çš„ç‰Œï¼ˆUI é¡ºåºç´¢å¼•ï¼‰
    std::vector<int> indices;
    for (int i = 0; i < m_handPanels.size(); ++i) {
        if (m_handPanels[i]->isSelected())
            indices.push_back(i);
    }

    if (indices.empty()) {
        qDebug() << "æ²¡æœ‰é€‰æ‹©ç‰Œ";
        return;
    }

    clearLastPlay();  // é˜²æ­¢ä¸Šä¸€æ¬¡æ˜¾ç¤ºå’Œè¿™æ¬¡å åœ¨ä¸€èµ·

    // ğŸ”´ æ³¨æ„ï¼šè¿™é‡Œå‡è®¾ Game::PlayerDiscard è¿”å› boolï¼Œ
    // å¦‚æœä½ ç°åœ¨è¿˜æ˜¯ voidï¼Œå¯ä»¥å…ˆæ”¹æˆ bool ç‰ˆæœ¬ï¼›
    // æˆ–è€…ä¸´æ—¶æŠŠä¸‹é¢å½“æˆâ€œå¿…å®šæˆåŠŸâ€çš„ç”¨æ³•ã€‚
    bool ok = m_game.PlayerDiscard(indices);

    if (!ok) {
        qDebug() << "âŒ å‡ºç‰Œå¤±è´¥ï¼ˆä¸åˆæ³•æˆ–æ— æ³•å‹ä¸Šï¼‰";
        return;
    }

    // UI åˆ é™¤å‡ºæ‰çš„ç‰Œ
    applyPlayerDiscardToUI(indices);

    // æ›´æ–°ä¸­é—´â€œä¸Šå®¶å‡ºçš„ç‰Œâ€
    showLastPlay();

    updateAiRemainLabels();

    // Game å†…éƒ¨å·²è‡ªåŠ¨æŠŠè½®åˆ°ä¸‹ä¸€ä½ï¼ˆåŒ…æ‹¬ AI çš„ DiscardPhaseï¼‰ï¼Œæ­¤å¤„åªéœ€è¦ç»§ç»­å¤„ç†é˜¶æ®µ
    enterDiscardPhase();
}

void InGameScene::onPassClicked()
{
    qDebug() << "ç©å®¶é€‰æ‹©è¿‡ç‰Œ";
    m_game.PlayerPass();
    // è¿‡ç‰Œåï¼Œä¸Šå®¶å‡ºç‰ŒåŒºåŸŸä»æ˜¾ç¤ºä¸Šä¸€æ‰‹ï¼Œæ‰€ä»¥ä¸è°ƒç”¨ showLastPlay()
    enterDiscardPhase();
}

void InGameScene::onHintClicked()
{
    qDebug() << "ç©å®¶è¯·æ±‚æç¤º";

    // 1. å…ˆæŠŠå½“å‰å·²ç»å¼¹èµ·çš„ç‰Œå…¨éƒ¨æ”¾å›å»ï¼ˆæ¸…ç©º UI é€‰ä¸­çŠ¶æ€ï¼‰
    int baseY = height() - 150;   // å’Œ createPlayer0HandPanels / applyPlayerDiscardToUI é‡Œä¿æŒä¸€è‡´
    for (CardPanel* panel : std::as_const(m_handPanels))
    {
        if (panel->isSelected())
        {
            panel->setSelected(false);
            panel->move(panel->x(), baseY);   // å¾€ä¸‹æ”¾ 20 åƒç´ 
        }
    }

    // 2. è°ƒç”¨æ¸¸æˆé€»è¾‘ç”Ÿæˆæç¤ºï¼ˆä¼šæŠŠæç¤ºæ–¹æ¡ˆå†™è¿›ç©å®¶çš„ selectionï¼‰
    m_game.PlayerHint();

    // 3. è¯»å–ç©å®¶å½“å‰çš„ selection
    Player* human = m_game.GetPlayer(0);
    if (!human) {
        qDebug() << "Hint: human player is null?";
        return;
    }

    const CardGroup& sel = human->GetSelection();
    const std::set<int>& hintCards = sel.GetCards();

    if (hintCards.empty()) {
        qDebug() << "Hint: æ²¡æœ‰å¯ä»¥æç¤ºçš„ç‰Œï¼ˆå¯èƒ½æ˜¯å‹ä¸ä¸Šï¼Œå»ºè®®è¿‡ç‰Œï¼‰";
        return;
    }

    // 4. éå†æ‰‹ç‰Œé¢æ¿ï¼Œå‡¡æ˜¯ç‰Œå·åœ¨ hintCards é‡Œçš„ï¼Œå°±å¼¹èµ·æ¥
    for (CardPanel* panel : std::as_const(m_handPanels))
    {
        // è¿™é‡Œç”¨çš„æ˜¯ä½  CardPanel é‡Œçš„ cardId() æ¥å£
        int id = panel->cardId();
        if (hintCards.find(id) != hintCards.end())
        {
            if (!panel->isSelected())
            {
                panel->setSelected(true);
                panel->move(panel->x(), panel->y() - 20);  // å¾€ä¸Šå¼¹ 20 åƒç´ 
            }
        }
    }
}

void InGameScene::setStatusText(const QString &text)
{
    ui->label_status->setText(text);
}




